src/relayclient/KnownRelaysManager.ts

```diff
@@ -27,39 +33,54 @@ const activeManagerEvents = ['RelayServerRegistered', 'TransactionRelayed', 'Can
 
 export interface IKnownRelaysManager {
   refresh (): Promise<void>
+
   saveRelayFailure (lastErrorTime: number, relayManager: Address, relayUrl: string): void
-  getRelaysSortedForTransaction (gsnTransactionDetails: GsnTransactionDetails): Promise<RelayRegisteredEventInfo[]>
+
+  getRelaysSortedForTransaction (gsnTransactionDetails: GsnTransactionDetails): Promise<RelayRegisteredEventInfo[][]>
 }
 
 export default class KnownRelaysManager implements IKnownRelaysManager {
-  private relayFailures = new Map<string, RelayFailureInfo[]>()
-  private readonly activeRelays = new Set<RelayRegisteredEventInfo>()
   private readonly contractInteractor: ContractInteractor
   private readonly config: GSNConfig
   private readonly relayFilter: RelayFilter
   private readonly scoreCalculator: AsyncScoreCalculator
+  private readonly httpClient: HttpClient
 
   private latestScannedBlock: number = 0
+  private relayFailures = new Map<string, RelayFailureInfo[]>()
+
+  public readonly knownRelays: RelayRegisteredEventInfo[][] = []
 
-  constructor (contractInteractor: ContractInteractor, config: GSNConfig, relayFilter?: RelayFilter, scoreCalculator?: AsyncScoreCalculator) {
+  constructor (contractInteractor: ContractInteractor, httpClient: HttpClient, config: GSNConfig, relayFilter?: RelayFilter, scoreCalculator?: AsyncScoreCalculator) {
     this.config = config
+    this.httpClient = httpClient
     this.relayFilter = relayFilter ?? EmptyFilter
     this.scoreCalculator = scoreCalculator ?? DefaultRelayScore
     this.contractInteractor = contractInteractor
   }
 
   /**
-   * Iterates through all relevant logs emitted by GSN contracts
-   * initializes an array {@link activeRelays}
+   * Pings all preferred arrays to discover their manager, fee information.
```

Comment:
ping doesn't give you fee info: for preferred relays, you don't know (and don't care) about advertised fees: they are always first
(you only use the worker (and maybe manager) you get from ping)